numba_typing/type_annotations.py

```diff
@@ -0,0 +1,89 @@
+from inspect import signature
+from typing import get_type_hints, Union, TypeVar
+from itertools import product
+from copy import deepcopy
+
+
+def get_func_annotations(func):
+    """Get annotations and default values of the fuction parameters."""
+    sig = signature(func)
+    annotations = {}
+    defaults = {}
+
+    for name, param in sig.parameters.items():
+        if param.annotation == sig.empty:
+            raise SyntaxError(f'No annotation for parameter {name}')
+
+        annotations[name] = get_annotation_types(param.annotation)
+        if param.default != sig.empty:
+            defaults[name] = param.default
+
+    return annotations, defaults
+
+
+def get_cls_annotations(cls):
+    """Get annotations of class attributes."""
+    annotations = get_type_hints(cls)
+    for x in annotations:
+        annotations[x] = get_annotation_types(annotations[x])
+    return annotations, {}
+
+
+def get_annotation_types(annotation):
+    """Get types of passed annotation."""
+    try:
+        if annotation.__origin__ is Union:
+            return list(annotation.__args__)
+    except AttributeError:
+        pass
+
+    return [annotation, ]
+
+
+def product_annotations(annotations):
+    """Get all variants of annotations."""
+    annot = annotations[0]
+    vals = annotations[1]
+    list_of_annot = list(product(*annot.values()))
+    tvs = {}
```

Comment:
You could consider defaultdict for this case:
```suggestion
    typevars = defaultdict(list)
```
Maybe this variable is not needed at all.